examples/barchart.rs

```diff
@@ -13,286 +13,267 @@
 
-    frame.render_widget(barchart, top);
-    draw_bar_with_group_labels(frame, app, left);
-    draw_horizontal_bars(frame, app, right);
+    // Rotate the data to simulate a real-time update
+    fn update_data(&mut self) {
+        let value = self.data.pop().unwrap();
+        self.data.insert(0, value);
+    }
 }
 
-#[allow(clippy::cast_precision_loss)]
-fn create_groups<'a>(app: &'a App, combine_values_and_labels: bool) -> Vec<BarGroup<'a>> {
-    app.months
-        .iter()
-        .enumerate()
-        .map(|(i, &month)| {
-            let bars: Vec<Bar> = app
-                .companies
-                .iter()
-                .map(|c| {
-                    let mut bar = Bar::default()
-                        .value(c.revenue[i])
-                        .style(c.bar_style)
-                        .value_style(
-                            Style::default()
-                                .bg(c.bar_style.fg.unwrap())
-                                .fg(Color::Black),
-                        );
-
-                    if combine_values_and_labels {
-                        bar = bar.text_value(format!(
-                            "{} ({:.1} M)",
-                            c.label,
-                            (c.revenue[i] as f64) / 1000.
-                        ));
-                    } else {
-                        bar = bar
-                            .text_value(format!("{:.1}", (c.revenue[i] as f64) / 1000.))
-                            .label(c.label.into());
-                    }
-                    bar
-                })
-                .collect();
-            BarGroup::default()
-                .label(Line::from(month).centered())
-                .bars(&bars)
+/// Generate some random data for the main bar chart
+fn generate_main_barchart_data() -> Vec<Bar<'static>> {
+    let mut rng = rand::thread_rng();
+    (1..50)
+        .map(|index| {
+            Bar::default()
+                .label(format!("B{index:>02}").into())
+                .value(rng.gen_range(1..15))
         })
         .collect()
 }
 
-#[allow(clippy::cast_possible_truncation)]
-fn draw_bar_with_group_labels(f: &mut Frame, app: &App, area: Rect) {
-    const LEGEND_HEIGHT: u16 = 6;
+impl Widget for &App<'_> {
+    /// Render the application
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        use Constraint::{Percentage, Ratio};
+        let [top, bottom] = Layout::vertical([Ratio(1, 3), Ratio(2, 3)]).areas(area);
+        let [left, right] = Layout::horizontal([Percentage(50), Percentage(50)]).areas(bottom);
+
+        let left_legend_area = App::legend_area(left);
+        let right_legend_area = App::legend_area(right);
+
+        self.main_barchart().render(top, buf);
+        self.vertical_revenue_barchart().render(left, buf);
+        self.horizontal_revenue_barchart().render(right, buf);
+        self.legend().render(left_legend_area, buf);
+        self.legend().render(right_legend_area, buf);
```

Comment:
`buf` is the convention that we use throughout ratatui. I think changing that is a ship that sailed long ago, and not one to change in an example like this. Consistency is more important here than using a more full name I think. (I say this as someone that dislikes abbreviations generally)

If we did change this, I'd want to change it everywhere all at once, in a major release so that all our docs / website code / examples all use buffer.

Reply:
> I think changing that is a ship that sailed long ago

Why?

The name in the trait does not need to be the name in the implementation. Only the types need to match. Nothing is breaking by changing that.

> If we did change this, I'd want to change it everywhere all at once, in a major release so that all our docs / website code / examples all use buffer.

This requires way more effort than just doing it when visiting it at some place, changing it while there. Seems to be a high bar for something that is relatively simple and painless.

Reply:
> Why?

Because it's used as buf conventionally everywhere, in every widget and every example (221 usages in ratatui), in code that's built with ratatui (GitHub search - 1.5K files) There are 14 files that use buffer instead of buf as a function parameter name.

> Nothing is breaking by changing that.

This is not about whether or not it would be breaking, this is about not making a change that isn't necessary. Buf as an abbreviation is well accepted. I didn't like it to begin with either, but my heuristic for choosing what to write weights convention over opinion in this.

I could get behind changing this across the board to drive things towards a new convention more broadly, but not piecemeal like this just in one example (or even all of them). I do agree with you that buffer is the right name for this parameter generally.